Do not unnecessarily merge non overlapping L0 groups in query results

watson28 opened the pull request:
solves: https://github.com/spirit-labs/tektite/issues/187

purplefox commented on levels/level_manager.go:
```diff
@@ -371,21 +372,21 @@ func (lm *LevelManager) getOverlappingTables(keyStart []byte, keyEnd []byte, lev
 	for {
 		// We load the readable state from an atomic pointer, this state is updated atomically after apply changes have
 		// been applied to the LSM, this allows us to reduce contention between the read path and the write path.
 		readable := lm.readable.Load()
```
The idea behind readable state is we use a copy on write approach. The readable state is set when it's updated, and then query can read the last readable state without having to use a lock, which reduces contention.
This means the query results returned all must be built from the readable state, it's not correct to use another lock and get the L0 groups, as the current writable state won't correspond to the same state if an update has been done in between.

purplefox commented on levels/level_manager.go:
```diff
@@ -398,10 +399,23 @@ outer:
-				// Level 0 is overlapping
-				for _, table := range tables {
-					overlapping = append(overlapping, []QueryTableInfo{{ID: table.SSTableID, DeadVersions: table.DeadVersionRanges}})
+				tablesByL0Group := lm.groupTablesByProcessorID(tables)
```
This doesn't work, as it assembles query results based on a combination of the readable state, and the current writable state, protected by a RW lock. Using a lock also defeats the purpose of readable state which is to avoid locking. If a compaction or registration occurs, it will get a write lock which will block all readers causing queries to stall while that is happening.

purplefox replied:
I think what we need to do is to add everything that we need to create the query results to the readable state.

watson28 commented:
>This doesn't work, as it assembles query results based on a combination of the readable state, and the current writable state, protected by a RW lock. Using a lock also defeats the purpose of readable state which is to avoid locking. If a compaction or registration occurs, it will get a write lock which will block all readers causing queries to stall while that is happening.
I think what we need to do is to add everything that we need to create the query results to the readable state.

@purplefox Thanks for the explanation. I have removed the lock and added `sstProcessorMap` to the readable state. 

I think it could be a good idea to extract the code of query/reading from the `level_manager.go` into its own module so it's clear the dependencies used for generating query result and thus easily verify any lock contention. For instance, the current `QueryTablesInRange` function depends on `segmentCache` ( `QueryTablesInRange -> getOverlappingTables -> getSegment -> segmentCache.get`. And segmentCache control access with a mutex lock so I assume there is contention when registering new tables `applyRegistrations()`.

purplefox commented:
Hey @watson28 sorry for the slow reply to this, and thank you for your work on it!
Over the last few days I've been thinking of some changes to the level manager which simplifies things and make its easier to partition our write load to the LSM. One result of that is that processor groups will disappear, which sadly makes this work redundant.
